src/shell/tmpdir.c

```diff
@@ -0,0 +1,123 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "src/common/libutil/cleanup.h"
+
+#include "builtins.h"
+#include "internal.h"
+#include "info.h"
+
+static int mkdir_exist_ok (const char *path, bool quiet)
+{
+    if (mkdir (path, 0700) < 0) {
+        if (errno != EEXIST) {
+            if (!quiet)
+                shell_log_errno ("mkdir %s", path);
+            return -1;
+        }
+    }
+    return 0;
+}
+
+static int make_job_path (flux_shell_t *shell,
+                          const char *parent,
+                          char *buf,
+                          size_t size)
+{
+    size_t n;
+
+    n = snprintf (buf, size, "%s/jobtmp-%d-", parent, shell->info->shell_rank);
+    if (n >= size) {
+        errno = EOVERFLOW;
+        return -1;
+    }
+    if (flux_job_id_encode (shell->jobid, "f58", buf + n, size - n) < 0)
+        return -1;
+    return 0;
+}
+
+static int mkjobtmp_rundir (flux_shell_t *shell, char *buf, size_t size)
+{
+    const char *rundir;
+
+    if (shell->standalone
+        || !(rundir = flux_attr_get (shell->h, "rundir"))
+        || make_job_path (shell, rundir, buf, size) < 0
+        || mkdir_exist_ok (buf, true) < 0)
+        return -1;
+    return 0;
+}
+
+static int mkjobtmp_tmpdir (flux_shell_t *shell, char *buf, size_t size)
+{
+    const char *tmpdir = flux_shell_getenv (shell, "TMPDIR");
+
+    if (make_job_path (shell, tmpdir ? tmpdir : "/tmp", buf, size) < 0
+        || mkdir_exist_ok (buf, false) < 0)
+        return -1;
+    return 0;
+}
+
+static int mktmpdir (flux_shell_t *shell)
+{
+    const char *tmpdir = flux_shell_getenv (shell, "TMPDIR");
+
+    if (tmpdir && mkdir_exist_ok (tmpdir, false) < 0)
+        return -1;
+    return 0;
+}
+
+static int tmpdir_init (flux_plugin_t *p,
+                        const char *topic,
+                        flux_plugin_arg_t *args,
+                        void *data)
+{
+    flux_shell_t *shell = flux_plugin_get_shell (p);
+    char jobtmp[1024];
+
+    /* Ensure TMPDIR exists if it is set in job environment
+     */
+    if (mktmpdir (shell) < 0)
+        return -1;
+
+    /* Try to create jobtmp in broker rundir.
+     * Fall back to ${TMPDIR:-/tmp} if that fails (e.g. guest user).
+     */
+    if (mkjobtmp_rundir (shell, jobtmp, sizeof (jobtmp)) < 0
+        && mkjobtmp_tmpdir (shell, jobtmp, sizeof (jobtmp)) < 0) {
+        shell_log_errno ("error creating jobtmp");
```

Comment:
Suggestion: improve error message by referencing `FLUX_JOB_TMPDIR` (users won't know what a `jobtmp` is):
```suggestion
        shell_log_errno ("error creating FLUX_JOB_TMPDIR");
```

Also, would it be trivial to add a test to ensure jobs fail predictably if FLUX_JOB_TMPDIR can't be created (unset rundir, then set TMPDIR to an unwritable location)?

Finally, since failure to create the job tmpdir is meant to be a fatal error, I wonder if using `shell_die_errno(3)` would be more appropriate here. This will cause a job exception to be immediately generated with the `note` set to the error message here, instead of later after `shell_init` returns -1, in which case the exception note will be a generic `"shell_init: %s", strerror (errno)`.

Reply:
Great suggestions: I was able to add that test, and it seems like huge improvement to get the more descriptive exception note in the user's job record on failure.  Thanks!